mkrlwe/marshaler.go

```diff
@@ -0,0 +1,332 @@
+	d2 := make([]byte, 8)
+	binary.LittleEndian.PutUint64(d2, evalGal.PeerID)
+
+	if err1 != nil {
+		panic("Couldn't Marshall evaluation key")
+	}
+	data = make([]byte, len(d1)+8)
+	for i := 0; i < len(d1); i++ {
+		data[i] = d1[i]
+	}
+	for i := 0; i < len(d2); i++ {
+		data[len(d1)+i] = d2[i]
+	}
+
+	return data, nil
+}
+
+// UnmarshalBinary decodes a previously marshaled Galois Evaluation Key in the target MKEvalGalKey.
+func (evalGal *MKRotationKey) UnmarshalBinary(data []byte) (err error) {
+
+	evalKey, err := UnmarshalSwitchingKey(data[:len(data)-8])
+	pid := binary.LittleEndian.Uint64(data[len(data)-8:])
+
+	if err != nil {
+		panic("Couldn't UnMarshall evaluation key")
+	}
+	evalGal.Key = &evalKey
+	evalGal.PeerID = pid
+
+	return nil
+}
+
+// MarshalBinary encodes an evaluation key in a byte slice.
+func (evalKey *MKRelinearizationKey) MarshalBinary() (data []byte, err error) {
+
+	// marshal each component of evalKey
+	d1, err1 := MarshalSwitchingKey(evalKey.Key01)
+	d2, err2 := evalKey.Key2.MarshalBinary()
+	d3 := make([]byte, 8)
+	binary.LittleEndian.PutUint64(d3, evalKey.PeerID)
+	if err1 != nil || err2 != nil {
+		panic("Couldn't Marshall evaluation key")
+	}
+
+	// Merge the byte slices
+	d12 := MergeTwoByteSlices(d1, d2)
+	data = make([]byte, len(d12)+8)
+	for i := 0; i < len(d12); i++ {
+		data[i] = d12[i]
+	}
+	for i := 0; i < len(d3); i++ {
+		data[len(d12)+i] = d3[i]
+	}
+
+	return data, nil
+}
+
+// UnmarshalBinary decodes a previously marshaled evaluation key in the target MKEvaluationKey.
+func (evalKey *MKRelinearizationKey) UnmarshalBinary(data []byte) (err error) {
+
+	// Unmarshal the switchingKey first
+	separator := binary.LittleEndian.Uint64(data[:8])
+	swkSize := binary.LittleEndian.Uint64(data[8:16])
+	value := make([][2]*ring.Poly, swkSize/2)
+
+	swk := new(rlwe.SwitchingKey)
+	var pointer, inc uint64
+	pointer = 16
+	dpStartIndex := separator + 8
+
+	for j := uint64(0); j < swkSize/2; j++ {
+		for k := 0; k < 2; k++ {
+			value[j][k] = new(ring.Poly)
+			if inc, err = value[j][k].DecodePolyNew(data[pointer:dpStartIndex]); err != nil {
+				return err
+			}
+			pointer += inc
+
+		}
+	}
+	swk.Value = value
+	evalKey.Key01 = swk
+
+	// Unmarshal the decomposedPoly
+	k2 := new(MKDecomposedPoly)
+	dpSize := binary.LittleEndian.Uint64(data[dpStartIndex : dpStartIndex+8])
+	dpoly := make([]*ring.Poly, dpSize)
+	pointer = dpStartIndex
+
+	for i := range dpoly {
+
+		dpoly[i] = new(ring.Poly)
+
+		if inc, err = dpoly[i].DecodePolyNew(data[pointer : len(data)-8]); err != nil {
+			return err
+		}
+
+		pointer += inc
+	}
+	k2.Poly = dpoly
+	evalKey.Key2 = k2
+
+	// Unmarshal the peerId
+	pid := binary.LittleEndian.Uint64(data[len(data)-8:])
+	evalKey.PeerID = pid
+
+	return nil
+
+}
+
+// MarshalSwitchingKey encodes a switching in a byte slice.
+func MarshalSwitchingKey(swk *rlwe.SwitchingKey) (data []byte, err error) {
+	dataLen, numEl := GetSwkDataLen(swk, true)
+	data = make([]byte, dataLen+7)
```

Comment:
Not really sure if I am correct, but I think it should be `+ 8`, since `binary.LittleEndian.PutUint64` takes 8 bytes. Same on line 262.